### PR TITLE
#155769359 Reject requests with spoofed information

### LINKED ISSUE #17: Requests with spoofed information should be rejected

### PR DESCRIPTION
**Pull request by jonathankamau**
## Resolves #17
 Pivotal Tracker Stories - [#155769359](https://www.pivotaltracker.com/story/show/155769359) and [#155769077](https://www.pivotaltracker.com/story/show/155769077)

## Description (what problem you're fixing)
- Currently there is no way to validate the payload from the decoded token to ensure that some of the information is not missing and is legitimate. 
- Also the code in `auth.py` requires refactoring.

## Fix (what you did to fix it)
- ensure that the token signature is verified using a public key and missing information is flagged.
- The authorization decorator in `auth.py` has been refactored and cleaned up.
- Add utils folder
- Refactor imports to match new structure
- Add tests for authorization decorator

## How to test (describe how to test your PR)
- consume one of the endpoints of the API and note the return message.
- run tests using the command `pytest`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -103,6 +103,9 @@ ENV/
 .mypy_cache/
 
 .vscode/
-test_db.sqlite
+*.sqlite
+
+# pipenv files
+Pipfile*
```
Comment (LarryWachira) [inline .gitignore]: I think you could just leave these for someone that uses the `pipenv` workflow.

File: api/models.py [hunk 2]
```diff
@@ -95,7 +94,8 @@ class User(Base):
     """Models Users."""
 
     __tablename__ = 'users'
-    email = db.Column(db.String)
+    name= db.Column(db.String, nullable=False, unique=True)
```
Comment (LarryWachira) [inline api/models.py]: It's possible that there might be more than one Andelan with the same name.

File: api/utils/auth.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    user = User.query.get(uuid)
+
+    # save user to db if they haven't been saved yet
+    if not user:
+        user = User(
+            uuid=uuid, name=name, email=email, photo=photo
+        )
+
+        user.society = random.choice(Society.query.all())
```
Comment (LarryWachira) [inline api/utils/auth.py]: This should be omitted now.

File: api/utils/auth.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        user.save()
+
+    # set current user in flask global variable, g
+    user.roles = roles
+    g.current_user = user
```
Comment (LarryWachira) [inline api/utils/auth.py]: Add a new line at end of file as indicated by the red icon after `user`.

File: .gitignore (hunk 1 above)
Comment (jonathankamau) [reply .gitignore]: Done. I have removed them from gitignore.

File: api/models.py (hunk 2 above)
Comment (jonathankamau) [reply api/models.py]: Good point. I have removed the `unique` value

File: api/utils/auth.py (hunk 4 above)
Comment (jonathankamau) [reply api/utils/auth.py]: Done. New line added.

File: api/utils/auth.py (hunk 3 above)
Comment (jonathankamau) [reply api/utils/auth.py]: Done. I have removed this from the code.

**Pull request merged**